### DISCUSSION TITLE
Detection model db_resnset50 detects very badly the words located in the right extremity of the document.

### DISCUSSION THREAD

**Discussion by ghassenBtb** [mindee/doctr#1137] [General]
Hello,

I noticed that the pretrained db_resnet50 is very bad at detecting words that are in the right extremity of the document.
If a document image contains many words at the right extremity the performances are bad. If I rotate by 180 degrees the document to make these words at the left extremity, the models detects them very accurately.

Does anyone have an explanation for this ? I used db_resnet50 with a resize of 1024*1024 with preserve_aspect_ratio=False.

**Comment by TomekPro** [comment]
Hi @ghassenBtb,

I had the same problem. In my case helped just naive widening of the image by a little margin.
`img = cv2.copyMakeBorder(img, 0, 0, 0, int(img.shape[1]*0.03), cv2.BORDER_CONSTANT, value=[255,255,255])`

**Comment by TomekPro** [comment]
I dived a little bit deeper - it seems that the model (db_resnet50) is biased, usually there are less tokens on the right and sometime logit values get very low on the far end:
![image](https://user-images.githubusercontent.com/49069202/219691399-6320cbce-4a29-4fc2-a6b4-9b66827ab134.png)

However when I manually add some "TEST" words on the far end again it starts to work properly:
![image](https://user-images.githubusercontent.com/49069202/219691620-af0b9948-3f62-45b8-918c-b1f4662be15e.png)

The idea that I shared previously (with adding margin on the far) slightly deteriorates overall accuracy :/ 

@felixdittrich92 any idea? Maybe the model should be retrained?

**Comment by ghassenBtb** [comment]
Hi @TomekPro
Yes the pretrained detection model is indeed biased towards words that are on the left side of the image, since it has been pretrained on documents where there are rarely annotated words on the right extremity of the document.

I also tried to add padding on the right side. Detection gets better, but the recognition performance deteriorates since cropped words fonts are impacted by the previous padding (small dilation of text pixels)

@felixdittrich92  One way to solve this is to pretrain the detection model on your same private internal dataset while adding upside-down rotation augmentation, so that the model avoid being biased toward words on the left side of the image. This should be easy to implement and it will allow to improve the overall performance of the doctr ocr :)

**Reply by felixdittrich92** [reply]
Hi @ghassenBtb :) 
I am not part of the mindee team i have only added a lot of stuff in my spare time so unfortunally  i am not able to trigger a training period 😅 (@ghassenBtb feel free to open a PR if you want to implement it) friendly ping @jonathanMindee @charlesmindee about the retrain topic

**Comment by TomekPro** [comment]
Hi, brute force but effective workaround for now is adding light grey lines on the right edge.Such lines are strong enough to make detection model recognize words on the right end and still so light that recognition model won't recognize it as text.
```
    for y in np.arange(0, img.shape[0], 300):
        img = cv2.putText(img=img, text='___', org=(int(img.shape[1]*0.97), y), fontFace=cv2.FONT_HERSHEY_TRIPLEX, fontScale=1, color=(240, 240, 240),thickness=5)
```
![image](https://user-images.githubusercontent.com/49069202/220060704-2e453541-e560-4b60-80ce-f120bedd9f23.png)

**Reply by DrRSatzteil** [reply]
I analyse pdf files so I had to write my own code but it did improve the detection performance a lot! Thank you again :)

**Reply by DrRSatzteil** [reply]
Actually I doubled the number of lines now which in fact increased the accuracy even further! Before that I still had some missing symbols on the right of the document.

**Reply by ghassenBtb** [reply]
Thanks for sharing, it did improve detection :)
Also, there is a newer version of db_resnet_50 which was apparently pretrained on documents using image rotation augmentations. You can try it out with the trick of adding light grey lines on the right edge, it might enhance the overall performance.
You can find information about the model here: https://github.com/mindee/doctr/blob/main/doctr/models/detection/differentiable_binarization/pytorch.py.
The model is named `db_resnet50_rotation` :
```
    "db_resnet50_rotation": {
        "input_shape": (3, 1024, 1024),
        "mean": (0.798, 0.785, 0.772),
        "std": (0.264, 0.2749, 0.287),
        "url": "https://doctr-static.mindee.com/models?id=v0.4.1/db_resnet50-1138863a.pt&src=0",
    }
```

**Reply by TomekPro** [reply]
Hi,
@DrRSatzteil great to hear that, however in my case there was no difference so I chose to modify img less.

@ghassenBtb Yes, I checked `db_resnet50_rotation` and in my case it deteriorated final BLEU on validation set by 6pp :/

**Reply by DrRSatzteil** [reply]
@TomekPro I think in the end I have a similar amount of stripes in my files now compared to yours when I looks your screenshot again. Since I painted this on a non rasterised image (a pdf) my document height is just not measured in pixels which is why steps of 300 height units resulted in less stripes for me.

Oh and I just paint black stripes, works as well. The user will never see this anyway so I don't care too much as long as the quality is improved 😊
